driver/src/orderbook/streamed/balance.rs

```diff
@@ -2,63 +2,104 @@ use super::*;
 use error::Error;
 use serde::{Deserialize, Serialize};
 
+/// The balance of a token for a user.
 #[derive(Clone, Copy, Debug, Default, Deserialize, Serialize)]
 pub struct Balance {
     balance: U256,
-    pub pending_deposit: Option<Flux>,
-    pub pending_withdraw: Option<Flux>,
+    deposit: Flux,
+    withdraw: Flux,
 }
 
 impl Balance {
-    pub fn deposit(&mut self, new_deposit: Flux, current_batch_id: BatchId) {
-        self.update_deposit_balance(current_batch_id);
-        match self.pending_deposit.as_mut() {
-            Some(deposit) => {
-                deposit.amount += new_deposit.amount;
-                deposit.batch_id = new_deposit.batch_id
-            }
-            None => self.pending_deposit = Some(new_deposit),
-        }
+    pub fn deposit(
+        &mut self,
+        amount: U256,
+        deposit_batch_id: BatchId,
+        current_batch_id: BatchId,
+    ) -> Result<(), Error> {
+        self.apply_existing_deposit(current_batch_id)?;
+        // Works like in the smart contract: If there is an existing deposit we override the
+        // batch id and add to the amount. If there is no existing deposit then amount is already 0.
```

Comment:
@fleupold 
> Do you need current_batch_id here? Couldn't we just compare `self.deposit.batch_id` to `deposit_batch_id`?
> 
> Otherwise I feel the `self.deposit.batch_id = deposit_batch_id;` might override some batchId that is in between `deposit_batch_id` and `current_batch_id`

I changed this a little bit compared to when you originally commented maybe it is more clear now.